src/Grpc.AspNetCore.Server/GrpcEndpointRouteBuilderExtensions.cs

```diff
@@ -20,60 +20,94 @@
     {
+        /// <summary>
+        /// Maps incoming requests to the specified <see cref="TService"/> type.
+        /// </summary>
+        /// <typeparam name="TService">The service type to map requests to.</typeparam>
+        /// <param name="builder">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <returns>An <see cref="IEndpointConventionBuilder"/> for endpoints associated with the service.</returns>
         public static IEndpointConventionBuilder MapGrpcService<TService>(this IEndpointRouteBuilder builder) where TService : class
         {
             if (builder == null)
             {
                 throw new ArgumentNullException(nameof(builder));
             }
 
+            return builder.MapGrpcService<TService>(configureOptions: null);
+        }
+
+        /// <summary>
+        /// Maps incoming requests to the specified <see cref="TService"/> type.
+        /// </summary>
+        /// <typeparam name="TService">The service type to map requests to.</typeparam>
+        /// <param name="builder">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="configureOptions">A callback to configure binding options.</param>
+        /// <returns>An <see cref="IEndpointConventionBuilder"/> for endpoints associated with the service.</returns>
+        public static IEndpointConventionBuilder MapGrpcService<TService>(this IEndpointRouteBuilder builder, Action<GrpcBindingOptions<TService>> configureOptions) where TService : class
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
             ValidateServicesRegistered(builder.ServiceProvider);
 
-            var service = typeof(TService);
+            var options = new GrpcBindingOptions<TService>();
+            options.BindAction = ReflectionBind<TService>;
+            options.InvokerFactory = new ReflectionMethodInvokerFactory<TService>();
+
+            configureOptions?.Invoke(options);
+
+            var callHandlerFactory = builder.ServiceProvider.GetRequiredService<ServerCallHandlerFactory<TService>>();
+            var serviceBinder = new GrpcServiceBinder<TService>(builder, callHandlerFactory, options.InvokerFactory);
 
             try
             {
-                // TService is an implementation of the gRPC service. It ultimately derives from Foo.TServiceBase base class.
-                // We need to access the static BindService method on Foo which implicitly derives from Object.
-                var baseType = service.BaseType;
-                while (baseType?.BaseType?.BaseType != null)
-                {
-                    baseType = baseType.BaseType;
-                }
-
-                // We need to call Foo.BindService from the declaring type.
-                var declaringType = baseType?.DeclaringType;
+                options.BindAction(serviceBinder, null);
```

Comment:
Is BindAction ever called with a non-null TService? If not, why have the second argument?

Reply:
Because code gen outputs a second argument that we don't use. We could make BindAction only take ServiceBinderBase but user code would go from:

```cs
builder.MapGrpcService<GreeterService>(options => options.BindAction = Greet.Greeter.BindService);
```

To:

```cs
builder.MapGrpcService<GreeterService>(options => options.BindAction = (binder) => Greet.Greeter.BindService(binder, null));
```